test/TritonGPU/combine.mlir

```diff
@@ -3827,15 +3827,19 @@ module attributes {"ttg.num-warps" = 4 : i32, "ttg.target" = "cuda:80"} {
 
 // -----
 
+// CHECK-DAG: [[BLOCKED_OUT:#.*]] = #ttg.blocked<{sizePerThread = [1, 1, 2]
+// CHECK-DAG: [[BLOCKED_JOIN:#.*]] = #ttg.blocked<{sizePerThread = [1, 2, 2]
+// CHECK-DAG: [[BLOCKED_IN:#.*]] = #ttg.blocked<{sizePerThread = [1, 2]
 #blocked = #ttg.blocked<{sizePerThread = [1, 1, 2], threadsPerWarp = [2, 16, 1], warpsPerCTA = [1, 1, 1], order = [2, 1, 0]}>
 #blocked1 = #ttg.blocked<{sizePerThread = [1, 1], threadsPerWarp = [2, 16], warpsPerCTA = [1, 1], order = [1, 0]}>
 #blocked2 = #ttg.blocked<{sizePerThread = [1, 2], threadsPerWarp = [2, 16], warpsPerCTA = [1, 1], order = [1, 0]}>
 module attributes {"ttg.num-warps" = 1 : i32, ttg.target = "cuda:80"} {
-  // CHECK-LABEL: join_forward
   tt.func @join_forward(%arg0: tensor<2x16xf32, #blocked2>) -> tensor<2x16x2xf32, #blocked> {
-    // CHECK: tt.join
-    // CHECK: ttg.convert_layout
-    // CHECK: tt.return
+    // Note this is a free conversion
+    // CHECK: [[SLICE:%.*]] = ttg.convert_layout %arg0 : tensor<2x16xf32, [[BLOCKED_IN]]> -> tensor<2x16xf32, #ttg.slice<{dim = 2, parent = [[BLOCKED_JOIN]]}>>
+    // CHECK: [[JOIN:%.*]] = tt.join [[SLICE]], [[SLICE]] : tensor<2x16xf32, #ttg.slice<{dim = 2, parent = [[BLOCKED_JOIN]]}>> -> tensor<2x16x2xf32, [[BLOCKED_JOIN]]>
+    // CHECK: [[RES:%.*]] = ttg.convert_layout [[JOIN]] : tensor<2x16x2xf32, [[BLOCKED_JOIN]]> -> tensor<2x16x2xf32, [[BLOCKED_OUT]]
+    // CHECK: tt.return [[RES]]
```

Comment:
do you understand why we end up with 2 convert layout here?

Reply:
I looked into it and it's coming from the fact that layout propagation uses `inferDstEncoding`, then `LayoutPropagation::cloneElementwise` uses `inferSrcEncoding` and in this case we infer a different encoding going backward because we now infer a slice encoding.

I've included a fix in `cloneElementwise` to look at the existing input encodings and select any that match up after calling `inferDstEncoding`.